storage/scratchpad/src/sparse_merkle/mod.rs

```diff
@@ -528,6 +522,415 @@ where
         ret
     }
 
+    /// The batch version of `serial_update`. The function will return a Sparse Merkle Tree of the final state.
+    /// This method is optmized to avoid repetitive common path updating and parallelize the processing as much
+    /// as possible.
+    pub fn batch_update(
+        &self,
+        updates: Vec<(HashValue, &V)>,
+        proof_reader: &impl ProofRead<V>,
+    ) -> Result<Self, UpdateError> {
+        // Flatten, dedup and sort the updates with a btree map since the updates between different
+        // versions may overlap on the same address in which case the latter always overwrites.
+        let kvs = updates
+            .into_iter()
+            .collect::<BTreeMap<_, _>>()
+            .into_iter()
+            .collect::<Vec<_>>();
+
+        let current_root = self.root_weak();
+        if kvs.is_empty() {
+            Ok(self.clone())
+        } else {
+            let root = Self::batch_update_subtree(
+                current_root,
+                &kvs[..],
+                /* depth = */ 0,
+                proof_reader,
+                /* is_generated_from_proofs = */ false,
+            )?;
+            Ok(Self::new_with_base(root, self))
+        }
+    }
+
+    /// Return the index of the first bit that is 1 at the given depth when kvs are lexicographically sorted.
+    fn partition<T>(kvs: &[(HashValue, &T)], depth: usize) -> usize {
+        // Binary search for the cut-off point where the bit at this depth turns from 0 to 1.
```

Comment:
could we not use binary_search_by function?

Reply:
partition_point seems even better https://github.com/rust-lang/rust/issues/73831

Reply:
partition_point will be stabilized in 1.52, at which time we can replace a bunch of hand written binary search in codebase. but it's not available now.

Reply:
Got it, thanks